Fix constraints in fixtures

HavokInspiration opened the pull request:
Currently, when you load your fixtures, you have to carefully order them so constraints are dealt with correctly to avoid some db drivers complaining about constraint integrity violation.
This PR aims to correct that by allowing you to put your fixtures and whatever order.

The same idea than with the migration plugin was applied : first, all tables are created and only after that are constraints added. Same thing goes for a truncate or a drop : first constraints are dropped and then the truncate or drop is done.

Refs #6323

HavokInspiration committed:
Implements a way of dealing with constraints in the fixtures

Fixtures are first created without the foreign keys. Once all tables have been created, constraints are added. This prevents errors that can happend when referencing non existant columns. Before being truncated, tables have their constraints dropped to prevent constraint violation.

HavokInspiration committed:
Fix MySQL build

HavokInspiration committed:
Add support for FK constraints management for fixtures to the Postgres server

HavokInspiration committed:
Add support for FK constraints management for fixtures to the SqlServer db server

HavokInspiration committed:
Add new TestFixture methods to the FixtureInterface

HavokInspiration committed:
Add the methods added to the schema dialect to the BaseSchema abstract class

HavokInspiration committed:
Quotes the table name for add / drop constraints operation and add tests to those methods

HavokInspiration committed:
Fix SqlServerSchema tests

HavokInspiration committed:
Add support for FK constraints management for fixtures to the SQLite db server

HavokInspiration committed:
Reorder fixtures to prove that the order does not matter anymore

HavokInspiration committed:
Fix tests for SqlServer

markstory commented on src/Database/Schema/SqliteSchema.php:
```diff
@@ -355,6 +355,100 @@ public function constraintSql(Table $table, $name)
+            $dropSql
+        ];
+
+        return $sql;
```
Oh SQLite..

HavokInspiration replied:
Tell me about it... FYI, there is a drop at the beginning and at the end because in the tests, I sometimes get a weird race condition (or that is was I suspect it is) where the table is locked and the last drop can't be performed because it is still busy doing the data copy. Which leads to a big mess because the next iteration of dealing with constraints raises errors (table can't be renamed because a table with the same name already exists and so on).
The only workaround I could find was to add a DROP as the first query to clear any leftovers.

I did some search and the proposed solution is to use ``closeCursor()`` when executing multiple queries in a row with SQLite, which is done, so I don't really understand that error.

And it gets weirder... I get that error from a specific test ([this one](https://github.com/cakephp/cakephp/pull/7503/files#diff-c22681210f8a3908b6e2b9cb94148ba1R482)) which is no different from the test right underneath which pass with no problem. Full error (there is a debug still lying around) can be found on [this build](https://travis-ci.org/cakephp/cakephp/jobs/83401317).

So if you guys have any idea...

markstory commented on src/Datasource/FixtureInterface.php:
```diff
@@ -48,6 +48,24 @@ public function drop(ConnectionInterface $db);
+     * @param \Cake\Datasource\ConnectionInterface $db An instance of the database into which the constraints will be dropped
+     * @return bool on success or if there are no constraints to drop, or false on failure
+     */
+    public function dropConstraints(ConnectionInterface $db);
```
These new methods need to be added to cakephp/elasticsearch before this merges.

HavokInspiration replied:
I can see what I can do about that. I might need some guidance though, I don't really know elasticsearch.

lorenzo replied:
The methods can be just empty. No need to implement anything there

HavokInspiration replied:
Ok, thanks.

markstory replied:
Another alternative would be to not add new methods to the interface and use `method_exists()` to check for these new methods.

markstory commented on src/TestSuite/Fixture/FixtureManager.php:
```diff
@@ -58,6 +57,13 @@ class FixtureManager
+     *
+     * @var array
+     */
+    protected $_processed = [];
```
:+1: This property was missing a formal definition.

markstory commented on src/TestSuite/Fixture/FixtureManager.php:
```diff
@@ -352,11 +376,17 @@ public function loadSingle($name, $db = null, $dropTables = true)
                 $db = ConnectionManager::get($fixture->connection());
             }
 
+            $fixture->dropConstraints($db);
```
In the case where you are loading a single fixture do you need to do constraints as a separate step?

HavokInspiration replied:
I'll have to check. But you might be right.

HavokInspiration replied:
Ah, I mis-read your comment.
Yes it is needed given that constraints are stored separately when the fixture schema is built from the ``$fields`` property : https://github.com/HavokInspiration/cakephp/blob/fixtures-order/src/TestSuite/Fixture/TestFixture.php#L171-L175

markstory commented on src/TestSuite/Fixture/TestFixture.php:
```diff
@@ -159,7 +166,11 @@ protected function _schemaFromFields()
         if (!empty($this->fields['_constraints'])) {
             foreach ($this->fields['_constraints'] as $name => $data) {
-                $this->_schema->addConstraint($name, $data);
+                if ($data['type'] !== 'foreign') {
```
Use the constant on Schema\Table?

HavokInspiration replied:
Yep, I missed that one.

markstory commented on src/TestSuite/Fixture/TestFixture.php:
```diff
@@ -80,6 +80,13 @@ class TestFixture implements FixtureInterface
+     *
+     * @var array
+     */
+    public $constraints = [];
```
Can we have a protected property instead of a public? Public properties always cause problems long term.

HavokInspiration replied:
Sure

markstory commented on tests/TestCase/Database/Schema/MysqlSchemaTest.php:
```diff
@@ -713,6 +713,120 @@ public function testIndexSql($name, $data, $expected)
+            ->addColumn('category_name', [
+                'type' => 'integer',
+                'null' => false
+            ])
```
Given that your test is focused on constraints you don't need to add all these columns.

markstory commented on tests/TestCase/Database/Schema/MysqlSchemaTest.php:
```diff
@@ -713,6 +713,120 @@ public function testIndexSql($name, $data, $expected)
+            ->addColumn('category_name', [
+                'type' => 'integer',
+                'null' => false
+            ])
```
Given that your test is focused on constraints you don't need to add all these columns.

HavokInspiration replied:
Actually they are needed. ``Table`` prevents constraints to be added if the the referenced columns don't exist.

markstory replied:
Ah good point, I didn't see category_name in the constraints.

markstory commented on tests/TestCase/Database/Schema/PostgresSchemaTest.php:
```diff
@@ -796,6 +796,120 @@ public function testConstraintSql($name, $data, $expected)
+            ->addColumn('category_name', [
+                'type' => 'integer',
+                'null' => false
+            ])
```
Given that your test is focused on constraints you don't need to add all these columns.

markstory commented on tests/TestCase/Database/Schema/SqliteSchemaTest.php:
```diff
@@ -545,6 +545,155 @@ public static function columnSqlProvider()
+)',
+            'INSERT INTO "posts"(author_id, category_id, category_name) SELECT author_id, category_id, category_name FROM "tmp_posts"',
+            'DROP TABLE IF EXISTS "tmp_posts"'
+        ];
```
If the scenario with multiple constraints works, I would expect a single constraint to work. You could probably do with only the one scenario.

lorenzo commented:
This is looking good!

HavokInspiration committed:
Use the \Cake\Database\Schema\Table constant instead of a plain text

HavokInspiration committed:
Convert the constraint property to a protected one and remove and unused property

HavokInspiration committed:
Clean up schema test cases by combining everything to one scenario

HavokInspiration commented:
I tried a new writing for SQLite, removing the leading ``DROP`` query (because that is quite ugly and leaving temporary tables in a db is not an option), wondering if the problem I have could come from the ``INSERT INTO (...) SELECT`` when there are no values for instance. It was passing locally, but Travis disagrees.

I'll try to see if I can produce another writing by using SQLite real temporary tables.

HavokInspiration commented:
So... I was able to have a failing build on a virtual machine at work which gave me the opportunity to study this a bit over lunch.
And my conclusions are unexpected : the problem seems to come from the ``ConnectionTest`` test file... only for PHP5.4 & 5.5. For the others versions, it fails.
If I remove the file [here is the build result](https://travis-ci.org/cakephp/cakephp/builds/83733500). 
With the file, [here is the build](https://travis-ci.org/cakephp/cakephp/builds/83546152).

If I load it last in the DatabaseSuite (after the directories Database and ORM), [here is the build](https://travis-ci.org/cakephp/cakephp/builds/83735362).

At first, I thought it was because some statement in ConnectionTest don't have a ``closeCursor()`` but tests I run with this in mind does not seem to be conclusive.

So to conclude, inconsistency all over the place.

If you have suggestions or idea where to look into, I'll happily try them.

HavokInspiration commented:
I'm not even sure that is a SQLite feature problem.
I run a 3.8.2 (5.5.9) on my local set up and tests pass.

Travis runs 
- 3.7.7.1 for PHP5.4 (fails)
- 3.8.7.2 for PHP5.5 (fails)
- 3.8.7.2 for PHP5.6 (pass sometimes)
- 3.8.10.2 for PHP7 (fails)

HavokInspiration commented:
Other tests : I reduced the DatabaseSuite to only one of the permutation.

[Build only without identifier quoting](https://travis-ci.org/cakephp/cakephp/builds/83775012)
[Build only with identifier quoting](https://travis-ci.org/cakephp/cakephp/builds/83777565)

The more I test stuff, the less I get what is going on.
Maybe this implementation is not the way to go.

markstory commented:
You might be tickling issues in the SQLite driver/SQLite. I know from debugkit that SQLite with `:memory:` can be very temperamental. Does using a file based database help at all?

HavokInspiration commented:
@markstory I tried with a file. Locally, that does not pass. And Travis seems to have the same opinion.

HavokInspiration commented:
I tried wrapping each constraints operations in a transaction after reading examples on the SQLite doc, but without any luck.

Out of curiosity, I changed the fixture used by the test that is failing.
The test now passes but I still get exceptions about database table lock thrown. And that fails appveyor.

I'm going to take a break from this and work on something else to see if I'll get an epiphany.

HavokInspiration commented:
Or we could not support adding / droping FK constraints on an already existing table for SQLite and have the method returns an empty array in order make them no-op.

markstory commented:
Having the fixture manager aware of various sql dialects/limitations feels like the wrong place for that logic. Perhaps the connection/fixture class can contain that logic using metadata from the drivers?

HavokInspiration committed:
Remove constraint support for SQLite

SQLite can not properly handle adding or dropping a constraint to an existing table.

HavokInspiration commented:
I pushed a commit where I removed support for adding / removing constraints to an existing table for SQLite. I'm not comfortable shipping something that we know might cause troubles in some cases.
If you are on board with this, I'll address the rest of the comments.

markstory commented on src/TestSuite/Fixture/TestFixture.php:
```diff
@@ -159,7 +169,11 @@ protected function _schemaFromFields()
         if (!empty($this->fields['_constraints'])) {
             foreach ($this->fields['_constraints'] as $name => $data) {
-                $this->_schema->addConstraint($name, $data);
+                if ($isSqlite || $data['type'] !== Table::CONSTRAINT_FOREIGN) {
```
Instead of a class check, could this be a method like `$connection->supportsDynamicConstraints()`? This would let other datasources/drivers implement it as necessary.

HavokInspiration replied:
That's a good idea.

HavokInspiration committed:
Fixture should be added to the insertionMap in any case dealt by the _setupTable() method

HavokInspiration committed:
Reorder operations done in a FixtureManager::loadSingle() call

HavokInspiration commented:
I addressed the remaining comments.
I'll try to do the necessary modifications to the elasticsearch repo tomorrow or over the weekend.

I did not find any tests for the ``FixtureManager::loadSingle()`` method. Should I add some ?

HavokInspiration committed:
Add a test to the FixtureManager::loadSingle() method

HavokInspiration commented:
I added a test for the ``FixtureManager::loadSingle()`` method.

markstory commented on src/TestSuite/Fixture/TestFixture.php:
```diff
@@ -285,6 +297,65 @@ public function insert(ConnectionInterface $db)
+                $db->execute($stmt)->closeCursor();
+            }
+        } catch (\Exception $e) {
+            debug($e);
```
Left over debug() call.

markstory replied:
We might not want to catch these errors. Letting them bubble up alerts the developer of an issue with their constraints/schema.

markstory commented:
Other than the debug() call this looks great. I would fix that issue but its on your fork.

HavokInspiration committed:
`addConstraints()` and `dropConstraints()` calls should let exceptions bubble up

HavokInspiration commented:
I pushed the change.
I'll try to take the habit to create working branches on the repo :smile:

lorenzo commented:
@HavokInspiration can you solve the merge issues?

HavokInspiration commented:
@lorenzo Done.

markstory commented:
Awesome work @HavokInspiration !

ADmad commented:
Using Cake 3.1.2 and debugkit 3.1.10 `MissingDatasourceConfigException` is thrown with message "The datasource configuration "test" was not found.", which is probably caused by changes done in this PR. I talked with @markstory on IRC and he seems to have an idea on what change is needed in debugkit to fix it.

In general though we should be a bit less accommodating when it comes to adding new features to bug fix release. Changes in this PR are non-trivial and would have been better to merge it to 3.2.
